File: pages/insurance_billing_page.py
"""
pages/insurance_billing_page.py
Page Object Model — Insurance & Billing tab.

Save pattern (confirmed from live DOM):
  - Toggle-only: Save Changes → API response
  - New Plan form: Save Plan (local) → Save Changes → API response
"""

import uuid
from playwright.sync_api import Page, Locator, expect
from pages.base_page import BasePage


class InsuranceBillingPage(BasePage):

    URL = "/settings?settingTab=Insurance+%26+Billing"

    CARD = {
        "coverage":         "Coverage",
        "membership_plans": "Membership Plans",
        "finance":          "Finance",
        "service_pricing":  "Service Pricing",
        "active_offers":    "Active Offers",
        "pricing_policy":   "Pricing Policy",
    }

    def __init__(self, page: Page) -> None:
        super().__init__(page)

    @staticmethod
    def unique(prefix: str = "Test") -> str:
        return f"{prefix}_{uuid.uuid4().hex[:6]}"

    # Navigation

    def navigate(self) -> None:
        self.page.goto(self.URL)
        self.page.wait_for_load_state("networkidle")

    def open_edit(self, card_name: str) -> None:
        edit_btn = self.page.locator(
            f"//h3[contains(text(),'{card_name}')]"
            "/ancestor::div[contains(@class,'flex')][2]"
            "//button[normalize-space()='Edit']"
        ).first
        edit_btn.scroll_into_view_if_needed()
        edit_btn.click()
        self.page.wait_for_timeout(500)

    # Modal

    @property
    def modal(self) -> Locator:
        return self.page.locator('[role="dialog"]')

    @property
    def save_button(self) -> Locator:
        return self.modal.get_by_role("button", name="Save Changes")

    @property
    def cancel_button(self) -> Locator:
        return self.modal.get_by_role("button", name="Cancel")

    @property
    def save_plan_button(self) -> Locator:
        return self.modal.get_by_role("button", name="Save Plan")

    @property
    def error(self) -> Locator:
        return self.modal.locator("p[id$='-error']").first

    def cancel(self) -> None:
        try:
            if self.cancel_button.is_visible():
                self.cancel_button.click()
                self.page.wait_for_timeout(300)
                discard = self.page.get_by_role("button", name="Discard")
                if discard.is_visible():
                    discard.click()
        except Exception:
            pass

    def save_and_assert_success(self) -> None:
        """Click Save Changes and verify via API response."""
        self.save_button.scroll_into_view_if_needed()
        with self.page.expect_response(
            lambda r: r.request.method in ("POST", "PUT", "PATCH")
                      and r.status in (200, 201, 204),
            timeout=10_000
        ):
            self.save_button.click()

    def save_plan_and_assert_success(self) -> None:
        """Save Plan (local state) then Save Changes (API)."""
        self.save_plan_button.scroll_into_view_if_needed()
        self.save_plan_button.click()
        self.page.wait_for_timeout(500)
        self.save_and_assert_success()

    def click_save(self) -> None:
        self.save_button.click(force=True)
        self.page.wait_for_timeout(500)

    def assert_delete_confirmation_shown(self) -> None:
        expect(self.page.get_by_text("cannot be undone")).to_be_visible()

    # Coverage

    @property
    def accept_all_toggle(self) -> Locator:
        return self.modal.locator('button[role="switch"]').first

    @property
    def add_custom_button(self) -> Locator:
        return self.modal.get_by_role("button", name="Add Custom")

    @property
    def insurance_name_input(self) -> Locator:
        return self.modal.locator('input[name="name"]')

    @property
    def payer_id_input(self) -> Locator:
        return self.modal.locator('input[name="payerId"]')

    @property
    def preventive_input(self) -> Locator:
        return self.modal.locator('input[name="preventiveCoverage"]')

    @property
    def basic_input(self) -> Locator:
        return self.modal.locator('input[name="basicCoverage"]')

    @property
    def major_input(self) -> Locator:
        return self.modal.locator('input[name="majorCoverage"]')

    @property
    def orthodontic_input(self) -> Locator:
        return self.modal.locator('input[name="orthodonticCoverage"]')

    @property
    def additional_notes(self) -> Locator:
        return self.modal.locator('textarea').first

    def fill_coverage_percentage(self, locator: Locator, value: str) -> None:
        locator.scroll_into_view_if_needed()
        locator.fill(value)
        locator.press("Tab")
        self.page.wait_for_timeout(300)

    def add_plan(self, name: str = None, payer_id: str = "12345") -> None:
        """Open Add Custom → fill required fields → Save Plan → Save Changes."""
        name = name or self.unique("Plan")
        self.add_custom_button.click()
        self.page.wait_for_timeout(500)
        self.smart_fill(self.insurance_name_input, name)
        self.insurance_name_input.press("Tab")
        self.page.wait_for_timeout(300)
        self.smart_fill(self.payer_id_input, payer_id)
        self.payer_id_input.press("Tab")
        self.page.wait_for_timeout(300)
        self.save_plan_and_assert_success()

    # Membership Plans

    @property
    def plan_name_input(self) -> Locator:
        return self.modal.locator('input[name="name"]')

    @property
    def annual_fee_input(self) -> Locator:
        return self.modal.locator('input[name="annualFee"]')

    @property
    def discount_percentage_input(self) -> Locator:
        return self.modal.locator('input[name="discountPercentage"]')

    # Finance

    @property
    def provider_name_input(self) -> Locator:
        return self.modal.locator('input[placeholder*="CareCredit"]')

    @property
    def provider_description(self) -> Locator:
        return self.modal.locator('textarea[placeholder*="Most popular"]')

    @property
    def provider_apr(self) -> Locator:
        return self.modal.locator('input[placeholder="26.99"]')

    @property
    def provider_key_features(self) -> Locator:
        return self.modal.locator('textarea[placeholder*="No prepayment"]')

    @property
    def in_house_financing_toggle(self) -> Locator:
        return self.modal.locator('button[role="switch"]').last

    # Service Pricing

    @property
    def cdt_code_input(self) -> Locator:
        return self.modal.locator('input[name="cdtCode"]')

    @property
    def service_name_input(self) -> Locator:
        return self.modal.locator('input[name="serviceName"]')

    @property
    def service_price_input(self) -> Locator:
        return self.modal.locator('input[name="price"]')

    # Active Offers

    @property
    def offer_name_input(self) -> Locator:
        return self.modal.locator('input[name="name"]')

    @property
    def promotional_price_input(self) -> Locator:
        return self.modal.locator('input[name="price"]')

    @property
    def original_price_input(self) -> Locator:
        return self.modal.locator('input[name="originalPrice"]')

    @property
    def included_services_textarea(self) -> Locator:
        return self.modal.locator('textarea[placeholder*="services separated"]')

    @property
    def restrictions_textarea(self) -> Locator:
        return self.modal.locator('textarea[name="restrictions"]')

    @property
    def expiration_days_input(self) -> Locator:
        return self.modal.locator('input[name="expirationDays"]')

    # Pricing Policy

    @property
    def good_faith_toggle(self) -> Locator:
        return self.modal.locator('button[role="switch"]').first

    @property
    def custom_ai_script(self) -> Locator:
        return self.modal.locator('textarea').first

    def select_pricing_option(self, text: str) -> None:
        self.modal.get_by_text(text, exact=False).first.click()
        self.page.wait_for_timeout(300)

File: src/pages/InsuranceBillingPage.ts
import { Page, Locator, expect } from '@playwright/test';
import { BasePage } from './BasePage';

/**
 * InsuranceBillingPage
 *
 * URL: /settings?settingTab=Insurance+%26+Billing
 *
 * Inherits from BasePage:
 *   openEdit(), modal, saveButton, cancelButton, error,
 *   cancel(), saveAndAssertSuccess(), fill(), fillAndBlur(),
 *   waitForApiSuccess(), unique()
 *
 * Save patterns (confirmed from live DOM):
 *   Toggle-only:  saveAndAssertSuccess()
 *   New Plan:     addPlan() → Save Plan (local) → saveAndAssertSuccess()
 *
 * NOTE: input[name="name"] and input[name="price"] are shared across cards.
 * All locators scoped to this.modal — safe because modals never overlap.
 */
export class InsuranceBillingPage extends BasePage {
  static readonly URL = '/settings?settingTab=Insurance+%26+Billing';

  static readonly CARD = {
    coverage:        'Coverage',
    membershipPlans: 'Membership Plans',
    finance:         'Finance',
    servicePricing:  'Service Pricing',
    activeOffers:    'Active Offers',
    pricingPolicy:   'Pricing Policy',
  } as const;

  constructor(page: Page) {
    super(page);
  }

  override async navigate(): Promise<void> {
    await this.page.goto(InsuranceBillingPage.URL);
    await this.page.waitForLoadState('domcontentloaded');
  }

  // -------------------------------------------------------------------------
  // Coverage — Save Plan button (unique to this card's Add Custom flow)
  // -------------------------------------------------------------------------

  get savePlanButton(): Locator {
    return this.modal.getByRole('button', { name: 'Save Plan' });
  }

  async savePlanAndAssertSuccess(): Promise<void> {
    await this.savePlanButton.scrollIntoViewIfNeeded();
    await this.savePlanButton.click();
    await expect(this.savePlanButton).not.toBeVisible();
    await this.saveAndAssertSuccess();
  }

  async assertDeleteConfirmationShown(): Promise<void> {
    await expect(this.page.getByText('This action cannot be undone')).toBeVisible();
  }

  async cancelDelete(): Promise<void> {
    // Cancel button inside the Delete Plan dialog specifically
    await this.page.getByLabel('Delete Plan').getByRole('button', { name: 'Cancel' }).click();
  }

  // -------------------------------------------------------------------------
  // Coverage — Accepted Insurance Plans
  // -------------------------------------------------------------------------

  get acceptAllToggle(): Locator {
    return this.modal.locator('button[role="switch"]').first();
  }

  get addCustomButton(): Locator {
    return this.modal.getByRole('button', { name: 'Add Custom' });
  }

  get insuranceNameInput(): Locator {
    return this.modal.locator('input[name="name"]');
  }

  get payerIdInput(): Locator {
    return this.modal.locator('input[name="payerId"]');
  }

  get preventiveInput(): Locator {
    return this.modal.locator('input[name="preventiveCoverage"]');
  }

  get basicInput(): Locator {
    return this.modal.locator('input[name="basicCoverage"]');
  }

  get majorInput(): Locator {
    return this.modal.locator('input[name="majorCoverage"]');
  }

  get orthodonticInput(): Locator {
    return this.modal.locator('input[name="orthodonticCoverage"]');
  }

  get coverageNotes(): Locator {
    return this.modal.locator('textarea').first();
  }

  async addPlan(options: {
    name?: string;
    payerId?: string;
    preventive?: string;
    basic?: string;
    major?: string;
    orthodontic?: string;
  } = {}): Promise<{ name: string; payerId: string }> {
    const name    = options.name    ?? BasePage.unique('Plan');
    const payerId = options.payerId ?? '12345';

    await this.addCustomButton.click();
    await expect(this.insuranceNameInput).toBeVisible();

    await this.fillAndBlur(this.insuranceNameInput, name);
    await this.fillAndBlur(this.payerIdInput, payerId);

    if (options.preventive)  await this.fillAndBlur(this.preventiveInput,  options.preventive);
    if (options.basic)       await this.fillAndBlur(this.basicInput,       options.basic);
    if (options.major)       await this.fillAndBlur(this.majorInput,       options.major);
    if (options.orthodontic) await this.fillAndBlur(this.orthodonticInput, options.orthodontic);

    await this.savePlanAndAssertSuccess();
    return { name, payerId };
  }

  // -------------------------------------------------------------------------
  // Membership Plans
  // -------------------------------------------------------------------------

  get membershipNameInput(): Locator {
    return this.modal.getByRole('textbox', { name: 'Plan Name' });
  }

  get annualFeeInput(): Locator {
    return this.modal.getByRole('spinbutton', { name: 'Annual Fee' });
  }

  get discountPercentageInput(): Locator {
    return this.modal.getByRole('spinbutton', { name: 'Discount (%)' });
  }

  get updatePlanButton(): Locator {
    return this.modal.getByRole('button', { name: 'Update Plan' });
  }

  get addMembershipPlanButton(): Locator {
    return this.modal.getByRole('button', { name: 'Add Plan' });
  }

  get deletePlanButton(): Locator {
    // Confirmed via Playwright codegen — trash icon is 2nd button in plan row
    return this.modal.locator('.flex.gap-2.items-center > button:nth-child(2)').first();
  }

  /** Open the edit form for the first plan in the list */
  async openFirstPlanEdit(): Promise<void> {
    // The edit icon is an empty-text button — nth(5) targets first plan's edit
    await this.modal.getByRole('button').filter({ hasText: /^$/ }).nth(5).click();
    await expect(this.updatePlanButton).toBeVisible();
  }

  async updatePlanAndAssertSuccess(): Promise<void> {
    await this.updatePlanButton.click();
    await this.saveAndAssertSuccess();
  }

  async addMembershipPlanAndAssertSuccess(): Promise<void> {
    await this.addMembershipPlanButton.click();
    await this.saveAndAssertSuccess();
  }

  // -------------------------------------------------------------------------
  // Finance
  // -------------------------------------------------------------------------

  get addCustomProviderButton(): Locator {
    return this.modal.getByRole('button', { name: 'Add Custom' });
  }

  get addFinanceProviderButton(): Locator {
    return this.modal.getByRole('button', { name: 'Add Finance Provider' });
  }

  get providerNameInput(): Locator {
    return this.page.getByRole('textbox', { name: 'Provider Name' });
  }

  get providerDescription(): Locator {
    return this.page.getByRole('textbox', { name: 'Description' });
  }

  get providerWebsite(): Locator {
    return this.page.getByRole('textbox', { name: 'Website' });
  }

  get providerApr(): Locator {
    return this.page.getByRole('textbox', { name: 'APR / Interest Rate (%)' });
  }

  get providerPaymentTerms(): Locator {
    return this.page.getByRole('textbox', { name: 'Payment Terms' });
  }

  get providerLoanAmountRange(): Locator {
    return this.page.getByRole('textbox', { name: 'Loan Amount Range' });
  }

  get providerCreditRequirements(): Locator {
    return this.page.getByRole('textbox', { name: 'Credit Requirements' });
  }

  get providerKeyFeatures(): Locator {
    return this.page.getByRole('textbox', { name: 'Key Features' });
  }

  get inHouseFinancingToggle(): Locator {
    return this.modal.getByRole('switch').nth(1);
  }

  async addFinanceProviderAndAssertSuccess(): Promise<void> {
    await this.addFinanceProviderButton.click();
    await this.saveAndAssertSuccess();
  }

  // -------------------------------------------------------------------------
  // Service Pricing
  // -------------------------------------------------------------------------

  get cdtCodeInput(): Locator {
    return this.modal.locator('input[name="cdtCode"]');
  }

  get serviceNameInput(): Locator {
    return this.modal.locator('input[name="serviceName"]');
  }

  get servicePriceInput(): Locator {
    return this.modal.locator('input[name="price"]');
  }

  // -------------------------------------------------------------------------
  // Active Offers
  // -------------------------------------------------------------------------

  get offerNameInput(): Locator {
    return this.modal.locator('input[name="name"]');
  }

  get promotionalPriceInput(): Locator {
    return this.modal.locator('input[name="price"]');
  }

  get originalPriceInput(): Locator {
    return this.modal.locator('input[name="originalPrice"]');
  }

  get includedServicesTextarea(): Locator {
    return this.modal.locator('textarea[placeholder*="services separated"]');
  }

  get restrictionsTextarea(): Locator {
    return this.modal.locator('textarea[name="restrictions"]');
  }

  get expirationDaysInput(): Locator {
    return this.modal.locator('input[name="expirationDays"]');
  }

  // -------------------------------------------------------------------------
  // Pricing Policy
  // -------------------------------------------------------------------------

  get goodFaithToggle(): Locator {
    return this.modal.locator('button[role="switch"]').first();
  }

  get customAiScriptTextarea(): Locator {
    return this.modal.locator('textarea').first();
  }

  async selectPricingOption(text: string): Promise<void> {
    await this.modal.getByText(text, { exact: false }).first().click();
  }
}
